pest/src/stack.rs

```diff
@@ -0,0 +1,177 @@
+// pest. The Elegant Parser
+// Copyright (c) 2018 Dragoș Tiselice
+//
+// Licensed under the Apache License, Version 2.0
+// <LICENSE-APACHE or http://www.apache.org/licenses/LICENSE-2.0> or the MIT
+// license <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. All files in the project carrying such notice may not be copied,
+// modified, or distributed except according to those terms.
+
+/// Implementation of a `Stack` which maintains an log of `StackOp`s in order to rewind the stack
+/// to a previous state.
+#[derive(Debug)]
+pub struct Stack<T: Clone> {
+    ops: Vec<StackOp<T>>,
+    cache: Vec<T>,
+    snapshots: Vec<usize>
+}
+
+impl<T: Clone> Stack<T> {
+    /// Creates a new `Stack`.
+    pub fn new() -> Self {
+        Stack {
+            ops: vec![],
+            cache: vec![],
+            snapshots: vec![]
+        }
+    }
+
+    /// Returns `true` if the stack is currently empty.
+    pub fn is_empty(&self) -> bool {
+        self.cache.is_empty()
+    }
+
+    /// Returns the top-most `&T` in the `Stack`.
+    pub fn peek(&self) -> Option<&T> {
+        self.cache.last()
+    }
+
+    /// Pushes a `T` onto the `Stack`.
+    pub fn push(&mut self, elem: T) {
+        self.ops.push(StackOp::Push(elem.clone()));
```

Comment:
Cloning here might be costly. Again, hard to say before we see exactly how this will be implemented.

Reply:
Originally, the stack holds `Span` structs. I expected that that wouldn't necessarily have to change. `Span`s are pretty cheap to clone.